src/Microsoft.ML.TensorFlow/TensorflowTransform.cs

```diff
@@ -0,0 +1,392 @@
+                    ctx.SaveNonEmptyString(colName);
+
+                ctx.SaveNonEmptyString(_outputColName);
+            }
+
+            private TFSession LoadTFSession(byte[] modelBytes, string modelArg)
+            {
+                var graph = new TFGraph();
+                try
+                {
+                    graph.Import(modelBytes, "");
+                }
+                catch (Exception ex)
+                {
+                    if (!string.IsNullOrEmpty(modelArg))
+                        throw _host.Except($"TensorFlow exception triggered while loading model from '{modelArg}'");
+#pragma warning disable MSML_NoMessagesForLoadContext
+                    throw _host.ExceptDecode(ex, "Tensorflow exception triggered while loading model.");
+#pragma warning restore MSML_NoMessagesForLoadContext
+
+                }
+                return new TFSession(graph);
+            }
+
+            private ITensorValueGetter CreateTensorValueGetter<T>(IRow input, bool isVector, int colIndex, TFShape tfShape)
+            {
+                if (isVector)
+                    return new TensorValueGetterVec<T>(input, colIndex, tfShape);
+                else
+                    return new TensorValueGetter<T>(input, colIndex);
+            }
+
+            private ITensorValueGetter CreateTensorValueGetter(IRow input, TFDataType tfType, bool isVector, int colIndex, TFShape tfShape)
+            {
+                var type = TFTensor.TypeFromTensorType(tfType);
+                _host.AssertValue(type);
+                return Utils.MarshalInvoke(CreateTensorValueGetter<int>, type, input, isVector, colIndex, tfShape);
+            }
+
+            private ITensorValueGetter[] GetTensorValueGetters(IRow input)
+            {
+                var srcTensorGetters = new ITensorValueGetter[_inputColIndices.Length];
+                for (int j = 0; j < _inputColIndices.Length; j++)
+                {
+                    int colIndex = _inputColIndices[j];
+                    srcTensorGetters[j] = CreateTensorValueGetter(input, _tfInputTypes[j], _isVectorInput[j], colIndex, _tfInputShapes[j]);
+                }
+                return srcTensorGetters;
+            }
+
+            private Delegate MakeGetter(IRow input)
+            {
+                var type = TFTensor.TypeFromTensorType(_tfOutputType);
+                _host.Assert(type == _outputColType.ItemType.RawType);
+                return Utils.MarshalInvoke(MakeGetter<int>, type, input, _outputColType);
+            }
+
+            private Delegate MakeGetter<T>(IRow input, ColumnType columnType)
+            {
+                _host.AssertValue(input);
+                _host.Assert(typeof(T) == columnType.ItemType.RawType);
+
+                var srcTensorGetters = GetTensorValueGetters(input);
+
+                ValueGetter<VBuffer<T>> valuegetter = (ref VBuffer<T> dst) =>
+                {
+                    var runner = _session.GetRunner();
```

Comment:
>GetRunner [](start = 42, length = 9)

it's one runner per getter call, right? Runners cannot be reused between calls?

Reply:
Yes, the runner cannot be reused currently.

---
In reply to: [213757002](https://github.com/dotnet/machinelearning/pull/704#discussion_r213757002) [](ancestors = 213757002)